File: 432_All_O1_DataStructure.java
import java.util.*;

class Bucket{
    public int rank;
    public Set <String> keySet;
    public Bucket next;
    public Bucket prev;
    public Bucket(int cnt){
        rank = cnt;
        keySet = new HashSet();
        next = null;
        prev = null;
    }
}

public class AllOne {
    /** Initialize your data structure here. */
    private Map <String, Integer> keyValMap;
    private Map <Integer, Bucket> valKeyMap;
    private Bucket dummyHead;
    private Bucket dummyTail;

    public AllOne() {
        keyValMap = new HashMap();
        valKeyMap = new HashMap();
        dummyHead = new Bucket(0);
        dummyTail = new Bucket(0);
        dummyHead.next = dummyTail;
        dummyTail.prev = dummyHead;
        valKeyMap.put(0, dummyHead);
    }
    /** Inserts a new key <Key> with value 1. Or increments an existing key by 1. */
    public void inc(String key) {
        // update keyValMap
        keyValMap.put(key, keyValMap.getOrDefault(key, 0) + 1);
        int curVal = keyValMap.get(key);
        // update valKeyMap,  delete should after add
        addToBucket(key, curVal);
        if (curVal != 1){
            deleteFromBucket(key, curVal - 1);
        }
    }
    
    /** Decrements an existing key by 1. If Key's value is 1, remove it from the data structure. */
    public void dec(String key) {
        // update keyValMap
        if (! keyValMap.containsKey(key)){
            return;
        }
        keyValMap.put(key, keyValMap.get(key) - 1);
        int curVal = keyValMap.get(key);
        if (curVal == 0){
            keyValMap.remove(key);
        }
        // update valKeyMap, delete should after add
        if (curVal != 0){
            addToBucket(key, curVal);
        }
        deleteFromBucket(key, curVal + 1);
    }
    
    /** Returns one of the keys with maximal value. */
    public String getMinKey() {
        Bucket minBucket = dummyHead.next;
        if (minBucket.rank == 0){
            return "";
        }
        return minBucket.keySet.iterator().next();
    }
    
    /** Returns one of the keys with Minimal value. */
    public String getMaxKey() {
        Bucket maxBucket = dummyTail.prev;
        if (maxBucket.rank == 0){
            return "";
        }
        return maxBucket.keySet.iterator().next();        
    }

    private void deleteFromBucket(String key, int value){
        Bucket delBucket = valKeyMap.get(value);
        delBucket.keySet.remove(key);
        if (delBucket.keySet.size() == 0){
            delBucket.prev.next = delBucket.next;
            delBucket.next.prev = delBucket.prev;
            valKeyMap.remove(value);
        }
    }

    private void addToBucket(String key, int value){
        if(!valKeyMap.containsKey(value)){
            Bucket prevBucket = valKeyMap.get(value - 1);
            Bucket nextBucket = valKeyMap.get(value + 1);
            Bucket newBucket = new Bucket(value);
            valKeyMap.put(value, newBucket);
            if (prevBucket == null){
                // insert the new bucket before nextBucket
                newBucket.prev = nextBucket.prev;
                newBucket.next = nextBucket;
                nextBucket.prev = newBucket;
                newBucket.prev.next = newBucket;
            }else{
                // insert the new bucket after prevBucket
                newBucket.prev = prevBucket;
                newBucket.next = prevBucket.next;
                prevBucket.next = newBucket;
                newBucket.next.prev = newBucket;
            }
        }
        valKeyMap.get(value).keySet.add(key);
    }

    // public static void main(String[] args) {
    //     AllOne myds = new AllOne();
    //     myds.inc("a");
    //     myds.inc("b");
    //     myds.inc("b");
    //     myds.inc("b");
    //     myds.inc("b");
    //     myds.inc("b");
    //     myds.dec("c");
    //     myds.dec("b");
    //     // myds.dec("b");
    //     System.out.println(myds.getMaxKey() + " == b");
    //     System.out.println(myds.getMinKey() + " == a");

    // }

}
/**
 * Your AllOne object will be instantiated and called as such:
 * AllOne obj = new AllOne();
 * obj.inc(key);
 * obj.dec(key);
 * String param_3 = obj.getMaxKey();
 * String param_4 = obj.getMinKey();
 */

File: 432_All_O1_DataStructure.py
# Implement a data structure supporting the following operations:

# Inc(Key) - Inserts a new key with value 1. Or increments an existing key by 1. Key is guaranteed to be a non-empty string.
# Dec(Key) - If Key's value is 1, remove it from the data structure. Otherwise decrements an existing key by 1. If the key does not exist, this function does nothing. Key is guaranteed to be a non-empty string.
# GetMaxKey() - Returns one of the keys with maximal value. If no element exists, return an empty string "".
# GetMinKey() - Returns one of the keys with minimal value. If no element exists, return an empty string "".
# Challenge: Perform all these in O(1) time complexity.

from collections import defaultdict

class Bucket(object):
    def __init__(self, rank):
        self.rank = rank
        self.keySet = set()
        self.prev = None
        self.next = None

class AllOne(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self._keyValMap = defaultdict(int)
        self._valKeyMap = {}
        self._dummyHead = Bucket(0)
        self._dummyTail = Bucket(0)

        self._valKeyMap[0] = self._dummyHead
        self._dummyHead.next = self._dummyTail
        self._dummyTail.prev = self._dummyHead

    def inc(self, key):
        """
        Inserts a new key <Key> with value 1. Or increments an existing key by 1.
        :type key: str
        :rtype: void
        """
        # handle keyValMap
        self._keyValMap[key] += 1
        curVal = self._keyValMap[key]
        # handle valKeyMap
        self._addKeyToBucket(curVal, key)
        if curVal != 1:
            self._deleteKeyFromBucket(curVal - 1, key)
        

    def dec(self, key):
        """
        Decrements an existing key by 1. If Key's value is 1, remove it from the data structure.
        :type key: str
        :rtype: void
        """
        if key not in self._keyValMap:
            return
        # handle keyValMap
        self._keyValMap[key] -= 1
        curVal = self._keyValMap[key]
        if curVal == 0:
            self._keyValMap.pop(key)
        # handle valKeyMap
        if curVal != 0:
            self._addKeyToBucket(curVal, key)
        self._deleteKeyFromBucket(curVal + 1, key)
        
        

    def getMaxKey(self):
        """
        Returns one of the keys with maximal value.
        :rtype: str
        """
        maxBucket = self._dummyTail.prev
        if maxBucket.rank == 0:
            return ""
        return iter(maxBucket.keySet).next()

    def getMinKey(self):
        """
        Returns one of the keys with Minimal value.
        :rtype: str
        """
        minBucket = self._dummyHead.next
        if minBucket.rank == 0:
            return ""
        return iter(minBucket.keySet).next()

    def _addKeyToBucket(self, value, key):
        # if the Bucket doesnt exist, create the bucket first
        if value not in self._valKeyMap:
            newBucket = Bucket(value)
            newBucket.rank = value
            self._valKeyMap[value] = newBucket
            prevBucket = self._valKeyMap.get(value - 1, None)
            nextBucket = self._valKeyMap.get(value + 1, None)
            if prevBucket:
                newBucket.next = prevBucket.next
                newBucket.prev = prevBucket
                prevBucket.next = newBucket
                newBucket.next.prev = newBucket
            else:
                newBucket.next = nextBucket
                newBucket.prev = nextBucket.prev
                nextBucket.prev = newBucket
                newBucket.prev.next = newBucket
        self._valKeyMap[value].keySet.add(key)

    def _deleteKeyFromBucket(self, value, key):
        deleteBucket = self._valKeyMap[value]
        deleteBucket.keySet.remove(key)
        if len(deleteBucket.keySet) == 0:
            # handle valKeyMap
            self._valKeyMap.pop(value)
            # handle the linked List
            deleteBucket.prev.next = deleteBucket.next
            deleteBucket.next.prev = deleteBucket.prev

# if __name__ == '__main__':
#     myds = AllOne()
#     print myds.getMaxKey()
#     print myds.getMinKey()
#     myds.inc("inc")
#     print myds.getMaxKey()
#     print myds.getMinKey()


# Your AllOne object will be instantiated and called as such:
# obj = AllOne()
# obj.inc(key)
# obj.dec(key)
# param_3 = obj.getMaxKey()
# param_4 = obj.getMinKey()